File: bots/crime/cogs/welcome.py
import discord
from discord.ext import commands
from .utils.util import Emojis, Colors
from .utils.embedparser import to_object
from cogs.utilevents import blacklist

class welcome(commands.Cog):
    def __init__(self, bot: commands.Bot):
        self.bot = bot

    @commands.Cog.listener()
    async def on_ready(self):
        async with self.bot.db.cursor() as cursor: 
            await cursor.execute("CREATE TABLE IF NOT EXISTS welcome (guild INTEGER, message TEXT, channel INTEGER);") 
        await self.bot.db.commit()

    @commands.Cog.listener()
    async def on_member_join(self, member: discord.Member):
        async with self.bot.db.cursor() as cursor:
            await cursor.execute("SELECT * FROM welcome WHERE guild = {}".format(member.guild.id))
            check = await cursor.fetchone()
            if check is not None:
                msg = check[1]
                chan = check[2]
                channel = self.bot.get_channel(chan)
                user = member
                guild = member.guild
                z = msg.replace('{user}', f'{user}').replace('{user.name}', f'{user.name}').replace('{user.mention}', f'{user.mention}').replace('{user.avatar}', user.avatar.url).replace('{user.joined_at}', f'<t:{int(user.created_at.timestamp())}:R>').replace('{user.discriminator}', f'{user.discriminator}').replace('{guild.name}', f'{guild.name}').replace('{guild.count}', f'{guild.member_count}').replace('{guild.icon}', guild.icon.url).replace('{guild.id}', f'{guild.id}')
                x = await to_object(z)
                await channel.send(**x)
  
    @commands.group(aliases=["welc", "wlc"], invoke_without_command=True)
    @commands.cooldown(1, 5, commands.BucketType.user)
    @blacklist()
    async def welcome(self, ctx):
        if ctx.subcommand_passed is not None:
          embed = discord.Embed(color=0xf7f9f8, title="welcome", description="greet your users into your guild")
          embed.set_author(name=self.bot.user.name, icon_url=self.bot.user.avatar.url)
          embed.add_field(name="category", value="config")
          embed.add_field(name="commands", value=">>> ,welcome message\n,welcome channel\n,welcome config\n,welcome variables\n,welcome delete\n,welcome test", inline=False)
          embed.add_field(name="usage", value="```,welcome message hello {user}!```", inline=False)
          embed.set_footer(text=f"aliases: welc, wlc")
          embed.set_thumbnail(url=self.bot.user.display_avatar.url)
          embed.set_footer(text="powered by crime")
          await ctx.reply(embed=embed, mention_author=False)
  
  
    @welcome.command(help="configure a welcome message for your server", description="config")
    @commands.cooldown(1, 5, commands.BucketType.user)
    @blacklist()
    async def message(self, ctx, *, code=None):
        if not ctx.author.guild_permissions.manage_guild:
            embed = discord.Embed(color=Colors.default, description=f"{Emojis.warn} you are missing permissions **manage_guild**")
            await ctx.reply(embed=embed, mention_author=False)
            return
        embed=discord.Embed(description=f"set welcome message to `{code}`", color=Colors.default)
        async with self.bot.db.cursor() as cursor:
            await cursor.execute("SELECT * FROM welcome WHERE guild = {}".format(ctx.guild.id))
            check = await cursor.fetchone()
            if check is None:
                async with self.bot.db.cursor() as cursor:
                    await cursor.execute("INSERT INTO welcome (guild, message, channel) VALUES (?, ?, ?)", (ctx.guild.id, code, None))
                await self.bot.db.commit()
                await ctx.reply(embed=embed, mention_author=False)
            elif check is not None:
                async with self.bot.db.cursor() as cursor:
                    await cursor.execute("UPDATE welcome SET message = ? WHERE guild = ?", (code, ctx.guild.id))
                await self.bot.db.commit()
                await ctx.reply(embed=embed, mention_author=False)

    @welcome.command(help="configure a welcome channel for your server", description="config")
    @commands.cooldown(1, 5, commands.BucketType.user)
    @blacklist()
    async def channel(self, ctx, channel: discord.TextChannel=None):
        if not ctx.author.guild_permissions.manage_guild:
            embed = discord.Embed(color=Colors.default, description=f"{Emojis.warn} you are missing permissions **manage_guild**")
            await ctx.reply(embed=embed, mention_author=False)
            return
        embed=discord.Embed(description=f"set welcome channel to {channel.mention}", color=Colors.default)
        async with self.bot.db.cursor() as cursor:
            await cursor.execute("SELECT * FROM welcome WHERE guild = {}".format(ctx.guild.id))
            check = await cursor.fetchone()
            if check is None:
                async with self.bot.db.cursor() as cursor:
                    await cursor.execute("INSERT INTO welcome (guild, message, channel) VALUES (?, ?, ?", (ctx.guild.id, None, channel.id))
                await self.bot.db.commit()
                await ctx.reply(embed=embed, mention_author=False)
            elif check is not None:
                async with self.bot.db.cursor() as cursor:
                    await cursor.execute("UPDATE welcome SET channel = ? WHERE guild = ?", (channel.id, ctx.guild.id))
                await self.bot.db.commit()
                await ctx.reply(embed=embed, mention_author=False)


    @welcome.command(help="check your welcome settings for your server", description="config")
    @commands.cooldown(1, 5, commands.BucketType.user)
    @blacklist()
    async def config(self, ctx):   
      if not ctx.author.guild_permissions.manage_guild:
        embed = discord.Embed(color=Colors.default, description=f"{Emojis.warn} you are missing permissions **manage_guild**") 
        await ctx.reply(embed=embed, mention_author=False)
        return
        
      async with self.bot.db.cursor() as cursor:
        await cursor.execute("SELECT * FROM welcome WHERE guild = {}".format(ctx.guild.id))
        check = await cursor.fetchone()
        msg = check[1] or "welcome message not set"
        chan = f"<#{check[2]}>" or "welcome channel not set"
        embed=discord.Embed(title="welcome message config", color=Colors.default)
        embed.add_field(name="message", value=f"```{msg}```")
        embed.add_field(name="channel", value=f"{chan.mention}")
        embed.set_thumbnail(url=ctx.guild.icon.url)
        embed.set_footer(text="powered by crime")
        await ctx.reply(embed=embed, mention_author=False)

    @welcome.command(help="view the welcome variables", description="config")
    @commands.cooldown(1, 5, commands.BucketType.user)
    @blacklist()
    async def variables(self, ctx):   
      embed=discord.Embed(description="here is a list of variables used to greet your members", color=Colors.default)
      embed.add_field(name="member", value=">>> {user}\n{user.name}\n{user.mention}\n{user.avatar}\n{user.discriminator}\n{user.joined_at}")
      embed.add_field(name="guild", value=">>> {guild.name}\n{guild.count}\n{guild.icon}\n{guild.id}")
      embed.set_thumbnail(url=ctx.guild.icon.url)
      embed.set_footer(text="powered by crime")
      await ctx.reply(embed=embed, mention_author=False)

    @welcome.command(help="delete your welcome config", description="config")
    @commands.cooldown(1, 5, commands.BucketType.user)
    @blacklist()
    async def delete(self, ctx):   
      if not ctx.author.guild_permissions.manage_guild:
        embed = discord.Embed(color=Colors.default, description=f"{Emojis.warn} you are missing permissions **manage_guild**") 
        await ctx.reply(embed=embed, mention_author=False)
        return
      async with self.bot.db.cursor() as cursor:
        await cursor.execute("DELETE FROM welcome WHERE guild = {}".format(ctx.guild.id))
      await self.bot.db.commit()
      embed=discord.Embed(description=f"{Emojis.check} deleted the welcome config for *{ctx.guild.name}*", color=Colors.default)
      await ctx.reply(embed=embed, mention_author=False)


    @welcome.command(help="tests your welcome message", description="config")
    @commands.cooldown(1, 5, commands.BucketType.user)
    @blacklist()
    async def test(self, ctx):   
      if not ctx.author.guild_permissions.manage_guild:
        embed = discord.Embed(color=Colors.default, description=f"{Emojis.warn} you are missing permissions **manage_guild**") 
        await ctx.reply(embed=embed, mention_author=False)
        return
      async with self.bot.db.cursor() as cursor:
        await cursor.execute("SELECT * FROM welcome WHERE guild = {}".format(ctx.guild.id))
        check = await cursor.fetchone()
        if check is not None:
          msg = check[1]
          chan = check[2]
          channel = self.bot.get_channel(chan)
          user = ctx.author
          guild = ctx.guild
          z = msg.replace('{user}', f'{user}').replace('{user.name}', f'{user.name}').replace('{user.mention}', f'{user.mention}').replace('{user.avatar}', user.avatar.url).replace('{user.joined_at}', f'<t:{int(user.created_at.timestamp())}:R>').replace('{user.discriminator}', f'{user.discriminator}').replace('{guild.name}', f'{guild.name}').replace('{guild.count}', f'{guild.member_count}').replace('{guild.icon}', guild.icon.url).replace('{guild.id}', f'{guild.id}')
          x = await to_object(z)
          await channel.send(**x)
          await ctx.reply(f"<#{chan}>", mention_author=False)
        elif check is None:
          embed=discord.Embed(description=f"{Emojis.warn} welcome message isnt configured for *{ctx.guild.name}*", color=Colors.default)
          await ctx.reply(embed=embed, mention_author=False)


async def setup(bot):
    await bot.add_cog(welcome(bot))

File: bleed js/configuration/welcome.js
const db = require('quick.db')
const discord = require('discord.js')
const { default_prefix } = require("../../config.json");
const { color } = require("../../config.json");
const { approve } = require('../../emojis.json')
const { warn } = require('../../emojis.json')

module.exports = {
  name: "welcome",
  aliases: ['welc', 'wlc'],

  run: async (client, message, args) => {
    if (!message.member.hasPermission("MANAGE_GUILD")) return message.channel.send({ embed: { color: "efa23a", description: `${warn} ${message.author}: You're **missing** permission: \`manage_guild\`` } });

    let prefix = db.get(`prefix_${message.guild.id}`);
    if (prefix === null) { prefix = default_prefix; };
    if (message.author.bot) return;
    const embed = new discord.MessageEmbed()
      .setTitle(`**${prefix}welcome**`)
      .setDescription(`set up a welcome message when new members join`)
      .addField(`**subcommands**`, `${prefix}welcome channel ＊ set where to send welcome messages\n${prefix}welcome clear ＊ clear the welcome message\n${prefix}welcome message ＊ set the welcome message text\n${prefix}welcome variables ＊ see all the welcome message variables`)
      .addField(`**usage**`, `${prefix}welcome`)
      .addField(`**aliases**`, `welc, wlc`)
      .setColor(color)
    if (!args[0]) return message.channel.send(embed)
    if (args[0].toLowerCase() == 'message') {
      db.set(`welmessage_${message.guild.id}`, args.splice(1).join(' '))
      let wlcmsg = db.get(`welmessage_${message.guild.id}`)
      if (wlcmsg === null) {
        const setmsgembed = new discord.MessageEmbed()
          .setDescription(`${warn} ${message.author}: There is no **welcome message** set one with \`${prefix}welcome message\``)
          .setColor(`#efa23a`)
        return message.channel.send(setmsgembed)
      } else {
        const setembed = new discord.MessageEmbed()
          .setTitle("welcome message:")
          .setDescription(`\`\`\`` + wlcmsg + `\`\`\``)
          .setFooter(`see the varibles with ${prefix}welcome variables + test with ${prefix}welcome test\ndisable with ${prefix}welcome clear`)
          .setColor(color)
        return message.channel.send(setembed)
      }
    } else if (args[0].toLowerCase() == 'test') {
      let chx = db.get(`welchannel_${message.guild.id}`);
      if (chx === null) {
        return;
      }
      let welcome = db.get(`welmessage_${message.guild.id}`);
      if (welcome === null) {
        return;
      }
      welcome = welcome.replace('{user}', message.member);
      welcome = welcome.replace('{user.name}', message.author.username);
      welcome = welcome.replace('{user.tag}', message.author.tag);
      welcome = welcome.replace('{user.id}', message.author.id);
      welcome = welcome.replace('{membercount}', message.member.guild.memberCount);
      const ordinal = (message.guild.memberCount.toString().endsWith(1) && !message.guild.memberCount.toString().endsWith(11)) ? 'st' : (message.guild.memberCount.toString().endsWith(2) && !message.guild.memberCount.toString().endsWith(12)) ? 'nd' : (message.guild.memberCount.toString().endsWith(3) && !message.guild.memberCount.toString().endsWith(13)) ? 'rd' : 'th';
      welcome = welcome.replace('{membercount.ordinal}', message.member.guild.memberCount + ordinal);
      welcome = welcome.replace('{guild.name}', message.member.guild.name);
      welcome = welcome.replace('{guild.id}', message.member.guild.id);

      client.channels.cache.get(chx).send(welcome)
      message.channel.send({ embed: { color: "#a3eb7b", description: `${approve} ${message.author}: Successfully tested your **welcome message** in <#` + chx + `>` } })
      if (chx === null) {
        return message.channel.send({ embed: { color: "#efa23a", description: `${warn} ${message.author}: There is no **welcome channel** set for me to test this` } })
      }
    } else if (args[0].toLowerCase() == 'variables') {
      const member = message.author
      const ordinal = (message.guild.memberCount.toString().endsWith(1) && !message.guild.memberCount.toString().endsWith(11)) ? 'st' : (message.guild.memberCount.toString().endsWith(2) && !message.guild.memberCount.toString().endsWith(12)) ? 'nd' : (message.guild.memberCount.toString().endsWith(3) && !message.guild.memberCount.toString().endsWith(13)) ? 'rd' : 'th';
      const variablesembed = new discord.MessageEmbed()
        .setTitle(`welcome variables`)
        .setDescription(`\`{user}\` ＊ <@` + member + `>\n\`{user.name}\` ＊ ` + message.author.username + `\n\`{user.tag}\` ＊ ` + message.author.tag + `\n\`{user.id}\` ＊ ` + message.author.id + `\n\`{guild.name}\` ＊ ` + message.member.guild.name + `\n\`{guild.id}\` ＊ ` + message.member.guild.id + `\n\`{membercount}\` ＊ ` + message.member.guild.memberCount + `\n\`{membercount.ordinal}\` ＊ ` + message.member.guild.memberCount + ordinal)
        .setColor(color)
      message.channel.send(variablesembed)
    } else if (args[0].toLowerCase() == "channel") {
      let channel = message.mentions.channels.first()
      if (!channel) {
        const welcomechannel = new discord.MessageEmbed()
          .setTitle(`${prefix}welcome channel`)
          .setDescription(`set the channel 4 welcome msgs`)
          .addField(`**usage**`, `${prefix}welcome channel [#channel]`)
          .addField(`**aliases**`, `c, chan`)
          .setColor(color)
        return message.channel.send(welcomechannel)
      }
      db.set(`welchannel_${message.guild.id}`, channel.id)
      await message.channel.send(`Set the welcome channel to ${channel}`)
    } else if (args[0].toLowerCase() == "chan") {
      let channel = message.mentions.channels.first()
      if (!channel) {
        const welcomechannel = new discord.MessageEmbed()
          .setTitle(`${prefix}welcome channel`)
          .setDescription(`set the channel 4 welcome msgs`)
          .addField(`**usage**`, `${prefix}welcome channel [#channel]`)
          .addField(`**aliases**`, `c, chan`)
          .setColor(color)
        return message.channel.send(welcomechannel)
      }
      db.set(`welchannel_${message.guild.id}`, channel.id)
      await message.channel.send({ embed: { color: "#a3eb7b", description: `${approve} ${message.author}: Set the **welcome channel** to ${channel}` } })
    } else if (args[0].toLowerCase() == "c") {
      let channel = message.mentions.channels.first()
      if (!channel) {
        const welcomechannel = new discord.MessageEmbed()
          .setTitle(`${prefix}welcome channel`)
          .setDescription(`set the channel 4 welcome msgs`)
          .addField(`**usage**`, `${prefix}welcome channel [#channel]`)
          .addField(`**aliases**`, `c, chan`)
          .setColor(color)
        return message.channel.send(welcomechannel)
      }
      db.set(`welchannel_${message.guild.id}`, channel.id)
      await message.channel.send({ embed: { color: "#a3eb7b", description: `${approve} ${message.author}: Set the **welcome channel** to ${channel}` } })
    } else if (args[0].toLowerCase() == "clear") {
      db.delete(`welchannel_${message.guild.id}`)
      db.delete(`welmessage_${message.guild.id}`)
      return await message.channel.send({ embed: { color: "#a3eb7b", description: `${approve} ${message.author}: Successfully cleared the **welcome channel** & **message**` } })
    }
  }
}